### ISSUE TITLE
nghttpx cannot resolve pure IP addresses without resolver

### ISSUE THREAD

**Issue by thinred** [nghttp2/nghttp2#857]
While packaging nghttpx 1.20.0 for Debian, I noticed that systemd service fails to resolve pure ipv4 addresses for the frontend part, but succeeds for the backend part:

```
2017-03-25T22:54:58.939+01:00 1053 1053 927957f4 INFO (shrpx_config.cc:3640) Resolving backend address
2017-03-25T22:54:58.939+01:00 1053 1053 927957f4 INFO (shrpx_config.cc:3650) Host-path pattern: group 0: '/'
2017-03-25T22:54:58.939+01:00 1053 1053 927957f4 INFO (shrpx_config.cc:3653) group 0 -> 127.0.0.1:80, proto=http/1.1
2017-03-25T22:54:58.939+01:00 1053 1053 927957f4 INFO (shrpx_config.cc:3669) Catch-all pattern is group 0
2017-03-25T22:54:58.939+01:00 1053 1053 927957f4 INFO (shrpx_config.cc:3795) Address resolution for 127.0.0.1 succeeded: 127.0.0.1
2017-03-25T22:54:58.939+01:00 1053 1053 927957f4 INFO (shrpx_config.cc:3719) Resolved backend address: 127.0.0.1:80 -> 127.0.0.1:80
2017-03-25T22:54:58.939+01:00 1053 1053 927957f4 INFO (shrpx.cc:724) Unable to get IPv4 address for 127.0.0.1, port 3000: Name or service not known
```

It surprises me to see that even though the address does not require any resolution, it still fails.

It may make sense to unify the resolving of addresses in nghttpx. Moreover, it may make sense to first try to resolve with AI_NUMERICSERV and AI_NUMERICHOST, and *then* without them. It seems that getaddrinfo is rather stupid and tries to resolve addresses even if they don't require it.

Let me know what you think.

**Comment by thinred** [comment]
Related to this a bit: maybe the provided service file should depend on network-online.target instead of only network.target (see https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/).

**Comment by tatsuhiro-t** [comment]
Thank you for this bug report.
When I started nghttpx while networking service is not available, nghttpx could still start without problem.
nghttpx uses different user functions for frontend and backend name resolution, but they use getaddrinfo, and same flags (assuming that this is start-up, not backendconfig API).
Are they any system configuration settings to change the default behaviour?

As for network-online.target, I'm a bit concerned about the following sentence from the web site:

> It is strongly recommended not to pull in this target too liberally: for example network server software should generally not pull this in (since server software generally is happy to accept local connections even before any routable network interface is up), it's primary purpose is network client software that cannot operate without network.

Are there any other server software to pull this target?

**Comment by tatsuhiro-t** [comment]
I found that resolving 127.0.0.1 succeeded, but ::1 doesn't.  It is quite interesting.

**Comment by tatsuhiro-t** [comment]
Even more interesting, getaddrinfo failed to resolve ::1, even when I passed AI_NUMERICSERV | AI_NUMERICHOST.

**Comment by tatsuhiro-t** [comment]
That is AI_ADDRCONFIG to disable IPv6 facility, and it is not getaddrinfo's fault.  Sorry for the noise.

**Comment by thinred** [comment]
Hi Tatsuhiro,
I can reproduce it now even without rebooting my system. I stopped networking.service and network-manager.target and this is what I have in terms of interfaces (ip a):
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: wlp58s0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc pfifo_fast state DOWN group default qlen 1000
    link/ether 30:52:cb:80:9d:a1 brd ff:ff:ff:ff:ff:ff

Now when I restart nghttpx I get exactly what I posted before. It seems to be a more generic getaddrinfo issue? I'll investigate more.

**Comment by thinred** [comment]
I added some debugging and found the problem with python's getaddrinfo:
```
import socket
socket.getaddrinfo("127.0.0.1", "3000", socket.AF_INET, 0, 0, socket.AI_ADDRCONFIG)
```
this code fails when I have my networking stopped (sudo systemctl stop network-manager.service).
Interestingly it does not fail in two cases: 

- if you do not use AI_ADDRCONFIG
- if you do not use AF_INET (that's the case when you resolve backend address)

I think it makes sense, man getaddrinfo says: "If hints.ai_flags includes the AI_ADDRCONFIG flag, then IPv4 addresses are returned in the list pointed to by res only if the local system has at least one IPv4 address configured [...] The loopback  address  is not considered for this case as valid as a configured address. ". So when there is no address on non-loopback interfaces, the call fails. However, I don't understand why it does not fail if you don't specify ai_family! All returned addresses are from AF_INET:

```
In [4]: socket.getaddrinfo("127.0.0.1", "3000", 0, 0, 0, socket.AI_ADDRCONFIG)
Out[4]: 
[(2, 1, 6, '', ('127.0.0.1', 3000)),
 (2, 2, 17, '', ('127.0.0.1', 3000)),
 (2, 3, 0, '', ('127.0.0.1', 3000))]
```

**Comment by thinred** [comment]
I was able to "fix" it by completely dropping the AI_ADDRCONFIG flag. This flag depends on the current state of the networking configuration, in particular on the presence of addresses on non-loopback interfaces which is slightly problematic if they are configured late in the boot process.
I test it on my laptop, where I have no static IP address and I only get any address after I'm connected to Wifi. I think the original code would work completely fine if there would be an interface with a static IP address. I'm still not sure that AI_ADDRCONFIG is still a good idea, though.

**Comment by tatsuhiro-t** [comment]
Thank you.  Based on your observation and suggestion, I made a patch to fix this issue.  Could you try #858 patch?  It now retries getaddrinfo without AI_ADDRCONFIG if getaddrinfo with that flag failed.

**Comment by thinred** [comment]
Yes, this works for me, but I'm not sure this is the right way. Do you have a specific reason to use AI_ADDRCONFIG?

**Comment by tatsuhiro-t** [comment]
Thank you.  There are 2 reasons to keep using AI_ADDRCONFIG.  Firstly we have no bug report about it prior this report.  Secondly, its manual page suggests that it is useful in some situations:

> This flag is useful on, for example, IPv4-only  systems,
       to ensure that getaddrinfo() does not return IPv6 socket addresses that
       would always fail in connect(2) or bind(2).

It looks like AI_ADDRCONFIG is somewhat double edged sword.  I'm also not sure the change is the best thing.  But keeping backward compatible code base, as well as handling this specific situation is a good workaround for the time being.

**Comment by tatsuhiro-t** [comment]
#858 has been merged into master.  I make a new release with this patch.
I keep this thread open, so that we can discuss the ideal solution for this issue.
